python/lsst/pipe/tasks/characterizeImage.py

```diff
@@ -0,0 +1,493 @@
+        @param[in,out] kwargs  other keyword arguments for lsst.pipe.base.CmdLineTask
+        """
+        pipeBase.CmdLineTask.__init__(self, **kwargs)
+        if schema is None:
+            schema = SourceTable.makeMinimalSchema()
+        self.schema = schema
+        self.makeSubtask("installSimplePsf")
+        self.makeSubtask("repair")
+        self.makeSubtask("measurePsf", schema=self.schema)
+        if self.config.doMeasurePsf and self.measurePsf.usesMatches:
+            self.makeSubtask("astrometry")
+        self.makeSubtask("detectAndMeasure", schema=self.schema)
+        self._initialFrame = getDebugFrame(self._display, "frame") or 1
+        self._frame = self._initialFrame
+
+    @pipeBase.timeMethod
+    def run(self, dataRef, exposure=None, background=None, doUnpersist=True):
+        """!Characterize a science image and, if wanted, persist the results
+
+        This simply unpacks the exposure and passes it to the characterize method to do the work.
+
+        @param[in] dataRef: butler data reference for science exposure
+        @param[in,out] exposure  exposure to characterize (an lsst.afw.image.ExposureF or similar).
+            If None then unpersist from "postISRCCD".
+            The following changes are made, depending on the config:
+            - set psf to the measured PSF
+            - set apCorrMap to the measured aperture correction
+            - subtract background
+            - interpolate over cosmic rays
+            - update detection and cosmic ray mask planes
+        @param[in,out] background  model of background model already subtracted from exposure
+            (an lsst.afw.math.BackgroundList). May be None if no background has been subtracted,
+            which is typical for image characterization.
+            A refined background model is output.
+        @param[in] doUnpersist  if True the exposure is read from the repository
+            and the exposure and background arguments must be None;
+            if False the exposure must be provided.
+            True is intended for running as a command-line task, False for running as a subtask
+
+        @return same data as the characterize method
+        """
+        self._frame = self._initialFrame # reset debug display frame
+        self.log.info("Processing %s" % (dataRef.dataId))
+
+        if doUnpersist:
+            if exposure is not None or background is not None:
+                raise pipeBase.TaskError("doUnpersist true; exposure and background must be None")
+            exposure = dataRef.get("postISRCCD", immediate=True)
+        elif exposure is None:
+            raise pipeBase.TaskError("doUnpersist false; exposure must be provided")
+
+        exposureIdInfo = dataRef.get("expIdInfo")
+
+        charRes = self.characterize(
+            exposure = exposure,
+            exposureIdInfo = exposureIdInfo,
+            background = background,
+        )
+
+        if self.config.doWrite:
```

Comment:
Put the following in its own method, to keep this top-level method as slim as possible.